compiler/codegen/cg-expr.cpp

```diff
@@ -5311,31 +5342,50 @@ static void codegenPutGet(CallExpr* call, GenRet &ret) {
     //   localvar, locale, remote addr, get(4)==size, line, file
     //                                  get(4)==len  for array_get/put
     const char* fn;
+    std::vector<GenRet> args;
     TypeSymbol* dt;
     bool isget = true;
 
+    // we actually have 5 arguments with sublocale, but than line and file info
+    // are added.
+    bool hasSubloc = (call->numActuals() == 7);
```

Comment:
> I'm actually confused how this function is going to get called for the GPU case anyway. Don't we generate a call to chpl_comm_get_gpu, I don't see that in our primitives table, do we need to add it?

That would be one way of what I am doing here and I wouldn't be opposed. But there are primitives and runtime functions. For a GET, there are two runtime functions but one primitive. Which runtime call to make is decided based on the number of arguments.

I agree that the number of arguments is a shaky ground. I can take up Ben's suggestion of also checking for argument types to make things more robust. If we were to add the GPU version of the primitive instead, I'd be a bit worried of having code like:

```
if CHPL_LOCALE_MODEL=="gpu" then
  __primitive("chpl_comm_get_gpu");
else
  __primitive("chpl_comm_get");
```

So, I am for putting the ugliness in the compiler code instead of the module code. But as is (without further checks) this maybe way too ugly.

Are you game with tightening based on the argument types and providing meaningful error messages for wrong cases?

Reply:
I agree I don't like the idea of having these checks for `CHPL_LOCALE_MODEL` sprinkled throughout the module code. I assume there are some cases where (regardless of the state of CHPL_LOCALE_MODEL) you'd still want to just directly call `chpl_comm_get`?: otherwise couldn't we just keep what you have and move the check for `CHPL_LOCALE_MODEL==gpu` into `codegenPutGet` directly?

Anyway, assuming that wouldn't work (and even if did), I'd say at a minimum this comment needs to be updated because it's just inaccurate:

```
    // args are:
    //   localvar, locale, remote addr, get(4)==size, line, file
    //                                  get(4)==len  for array_get/put
```

But I like what Ben suggested: add some type checking checking code to identify which "overload" of the primitive is being used and add to that (in true defensive programming fashion) an assertion if we're fall into some new or invalid state.

Really, I think we should be doing this for all the primitives (it's good to be defensive and it's a lot more self-documenting than a comment that quickly gets forgotten about and falls out of date).  I don't think I'm going to be able to convince anyone (including myself) to update the existing primitives, but we can fix things as we go and a polymorphic-primitive like this one seems like a prime candidate for some checks like this.

Anyway, if it were me I'd probably timebox an hour, give it a shot, and if ended up being a hassle or the code ended up uglier, then I'd just update the comment and move on.

Reply:
> Really, I think we should be doing this for all the primitives (it's good to be defensive and it's a lot more self-documenting than a comment that quickly gets forgotten about and falls out of date). I don't think I'm going to be able to convince anyone (including myself) to update the existing primitives, but we can fix things as we go and a polymorphic-primitive like this one seems like a prime candidate for some checks like this.

I agree with this. I guess the main complication is that primitives could be created/consumed anytime during compilation before or after resolution. Maybe a Primitive type in the AST (as opposed to an enum and a flag on CallExpr) can provide a better infra for that support. dyno might be doing something of that sort.

I am currently considering sprinkling assertions in this function. i.e. whenever we pick a new `curArg` we check its type. I'll see what that turns out to be like.

Reply:
https://github.com/chapel-lang/chapel/pull/22349/commits/66704e7a957dc0af431b6a45d90682b73f2cea00 checks all 5 user-given arguments. Other 2 are line/file info. Does that look good?

Reply:
> I am currently considering sprinkling assertions in this function. i.e. whenever we pick a new curArg we check its type. I'll see what that turns out to be like.

Ok, thanks for doing that, it feels more safe to me.

I think this comment is still out of date, please update it if you agree:

```
    // args are:
    //   localvar, locale, remote addr, get(4)==size, line, file
    //                                  get(4)==len  for array_get/put
```

Reply:
2be1aa3f29